Nonexistent aggregation event Group by Field causes error: Invalid format: "(Empty Value)"

[HS 1357286194]
## Description:
When a `Group by Field` for an Aggregation event definition does not exist on any log messages queried by the event, the following exception occurs when the correlation processor runs:

```
Caused by: java.lang.IllegalArgumentException: Invalid format: "(Empty Value)"
	at org.joda.time.format.DateTimeFormatter.parseDateTime(DateTimeFormatter.java:945) ~[joda-time-2.10.6.jar:2.10.6]
```

<details>
<summary>Expand for full stacktrace:</summary>
ERROR: org.graylog.events.processor.EventProcessorExecutionJob - Event processor <aggregation-v1/63b880b6c03bed191e30bc94> failed to execute: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=63b880b6c03bed191e30bc94, title=Aggregation, description=, priority=2, alert=false, config=AggregationEventProcessorConfig{type=aggregation-v1, query=*, queryParameters=[], streams=[000000000000000000000001], groupBy=[kljhkl], series=[AggregationSeries{id=avg-jkhkhj, function=AVG, field=Optional[jkhkhj]}], conditions=Optional[AggregationConditions{expression=Optional[Greater{expr=>, left=NumberReference{expr=number-ref, ref=avg-jkhkhj}, right=NumberValue{expr=number, value=10.0}}]}], searchWithinMs=15000, executeEveryMs=15000}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=0, backlogSize=0}, notifications=[], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}]} (retry in 5000 ms)
org.graylog.events.processor.EventProcessorException: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=63b880b6c03bed191e30bc94, title=Aggregation, description=, priority=2, alert=false, config=AggregationEventProcessorConfig{type=aggregation-v1, query=*, queryParameters=[], streams=[000000000000000000000001], groupBy=[kljhkl], series=[AggregationSeries{id=avg-jkhkhj, function=AVG, field=Optional[jkhkhj]}], conditions=Optional[AggregationConditions{expression=Optional[Greater{expr=>, left=NumberReference{expr=number-ref, ref=avg-jkhkhj}, right=NumberValue{expr=number, value=10.0}}]}], searchWithinMs=15000, executeEveryMs=15000}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=0, backlogSize=0}, notifications=[], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}]}
	at org.graylog.events.processor.EventProcessorEngine.execute(EventProcessorEngine.java:105) ~[classes/:?]
	at org.graylog.events.processor.EventProcessorExecutionJob.execute(EventProcessorExecutionJob.java:115) ~[classes/:?]
	at org.graylog.scheduler.JobExecutionEngine.executeJob(JobExecutionEngine.java:173) ~[classes/:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$handleTrigger$2(JobExecutionEngine.java:151) ~[classes/:?]
	at com.codahale.metrics.Timer.time(Timer.java:151) ~[metrics-core-4.1.9.jar:4.1.9]
	at org.graylog.scheduler.JobExecutionEngine.handleTrigger(JobExecutionEngine.java:151) ~[classes/:?]
	at org.graylog.scheduler.JobExecutionEngine.lambda$execute$0(JobExecutionEngine.java:120) ~[classes/:?]
	at org.graylog.scheduler.worker.JobWorkerPool.lambda$execute$0(JobWorkerPool.java:121) ~[classes/:?]
	at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [metrics-core-4.1.9.jar:4.1.9]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at com.codahale.metrics.InstrumentedThreadFactory$InstrumentedRunnable.run(InstrumentedThreadFactory.java:66) [metrics-core-4.1.9.jar:4.1.9]
	at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.lang.IllegalArgumentException: Invalid format: "(Empty Value)"
	at org.joda.time.format.DateTimeFormatter.parseDateTime(DateTimeFormatter.java:945) ~[joda-time-2.10.6.jar:2.10.6]
	at org.joda.time.DateTime.parse(DateTime.java:160) ~[joda-time-2.10.6.jar:2.10.6]
	at org.joda.time.DateTime.parse(DateTime.java:149) ~[joda-time-2.10.6.jar:2.10.6]
	at org.graylog.events.processor.aggregation.PivotAggregationSearch.extractValues(PivotAggregationSearch.java:322) ~[classes/:?]
	at org.graylog.events.processor.aggregation.PivotAggregationSearch.doSearch(PivotAggregationSearch.java:161) ~[classes/:?]
	at org.graylog.events.processor.aggregation.AggregationEventProcessor.aggregatedSearch(AggregationEventProcessor.java:234) ~[classes/:?]
	at org.graylog.events.processor.aggregation.AggregationEventProcessor.createEvents(AggregationEventProcessor.java:124) ~[classes/:?]
	at org.graylog.events.processor.EventProcessorEngine.execute(EventProcessorEngine.java:91) ~[classes/:?]
</details>

## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
This issue can be reproduced by creating logs with the Random HTTP message generator input and creating the following aggregation event definition. Note the `field-does-not-exist` group by field.

![image](https://user-images.githubusercontent.com/3423655/211096597-c63fd097-c5c8-474b-a618-50d0d9bc4086.png)

## Workaround
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
Specify a `Group by Field` that exists on the log messages queried by the aggregation.

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

* Graylog Version: 5.0.2
* Java Version:
* Elasticsearch Version: 7.10.2
* MongoDB Version: 5.0.14 
* Operating System:
* Browser version:

Here's another related scenario, which we should deal with at the same time:
Define an aggregation event, grouped by 2 fields, one of which is valid and the other invalid (field does not occur in any message). This reliably throws the following type of exception:
```
2023-02-15 12:38:05,270 ERROR: org.graylog.events.processor.EventProcessorExecutionJob - Event processor <aggregation-v1/63ea64900e761c129a2a3d15> failed to execute: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=63ea64900e761c129a2a3d15, title=test, description=, priority=2, alert=false, config=AggregationEventProcessorConfig{type=aggregation-v1, query=, queryParameters=[], streams=[000000000000000000000001], groupBy=[a;lsdfkjadslfj, action], series=[AggregationSeries{id=avg-xxxxxx, function=AVG, field=Optional[xxxxxx]}], conditions=Optional[AggregationConditions{expression=Optional[Greater{expr=>, left=NumberReference{expr=number-ref, ref=avg-xxxxxx}, right=NumberValue{expr=number, value=2.0}}]}], searchWithinMs=15000, executeEveryMs=15000}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=300000, backlogSize=0}, notifications=[], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}]} (retry in 5000 ms)
org.graylog.events.processor.EventProcessorException: Couldn't create events for: EventDefinitionDto{scope=DEFAULT, id=63ea64900e761c129a2a3d15, title=test, description=, priority=2, alert=false, config=AggregationEventProcessorConfig{type=aggregation-v1, query=, queryParameters=[], streams=[000000000000000000000001], groupBy=[a;lsdfkjadslfj, action], series=[AggregationSeries{id=avg-xxxxxx, function=AVG, field=Optional[xxxxxx]}], conditions=Optional[AggregationConditions{expression=Optional[Greater{expr=>, left=NumberReference{expr=number-ref, ref=avg-xxxxxx}, right=NumberValue{expr=number, value=2.0}}]}], searchWithinMs=15000, executeEveryMs=15000}, fieldSpec={}, keySpec=[], notificationSettings=EventNotificationSettings{gracePeriodMs=300000, backlogSize=0}, notifications=[], storage=[Config{type=persist-to-streams-v1, streams=[000000000000000000000002]}]}
[...]
Caused by: java.lang.IndexOutOfBoundsException: index (1) must be less than size (1)
```

--- a/graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorEngine.java
+++ b/graylog2-server/src/main/java/org/graylog/events/processor/EventProcessorEngine.java
@@ -99,7 +99,7 @@ public class EventProcessorEngine {
             throw e;
         } catch (Exception e) {
             metrics.recordException(eventProcessor, definitionId);
-            LOG.error("Caught an unhandled exception while executing event processor <{}/{}/{}> - Make sure to modify the event processor to throw only EventProcessorExecutionException so we get more context!",
+            LOG.error("Caught an unhandled exception while executing event processor <{}/{}/{}> - Make sure to modify the event processor to throw only EventProcessorException so we get more context!",
                     definition.config().type(), definition.title(), definition.id(), e);
             // Since we don't know what kind of error this is, we play safe and make this a temporary error.
             throw new EventProcessorException("Couldn't create events for: " + definition.toString(), false, definition, e);

--- a/graylog2-server/src/main/java/org/graylog/events/processor/aggregation/AggregationEventProcessor.java
+++ b/graylog2-server/src/main/java/org/graylog/events/processor/aggregation/AggregationEventProcessor.java
@@ -270,7 +270,8 @@ public class AggregationEventProcessor implements EventProcessor {
     }
 
     @VisibleForTesting
-    ImmutableList<EventWithContext> eventsFromAggregationResult(EventFactory eventFactory, AggregationEventProcessorParameters parameters, AggregationResult result) {
+    ImmutableList<EventWithContext> eventsFromAggregationResult(EventFactory eventFactory, AggregationEventProcessorParameters parameters, AggregationResult result)
+            throws EventProcessorException {
         final ImmutableList.Builder<EventWithContext> eventsWithContext = ImmutableList.builder();
         final Set<String> sourceStreams = eventStreamService.buildEventSourceStreams(getStreams(parameters),
                 result.sourceStreams());
@@ -312,7 +313,13 @@ public class AggregationEventProcessor implements EventProcessor {
             //   application_name=sshd
             //   username=jane
             for (int i = 0; i < config.groupBy().size(); i++) {
-                fields.put(config.groupBy().get(i), keyResult.key().get(i));
+                try {
+                    fields.put(config.groupBy().get(i), keyResult.key().get(i));
+                } catch (IndexOutOfBoundsException e) {
+                    throw new EventProcessorException(
+                            "Couldn't create events for: " + eventDefinition.title() + " (possibly due to non-existing grouping fields)",
+                            false, eventDefinition.id(), eventDefinition, e);
+                }
             }
 
             // Group By fields need to be saved on the event so they are available to the subsequent notification events

--- a/graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java
+++ b/graylog2-server/src/main/java/org/graylog/events/processor/aggregation/PivotAggregationSearch.java
@@ -158,7 +158,7 @@ public class PivotAggregationSearch implements AggregationSearch {
         final PivotResult streamsResult = (PivotResult) streamQueryResult.searchTypes().get(STREAMS_PIVOT_ID);
 
         return AggregationResult.builder()
-                .keyResults(extractValues(pivotResult))
+                .keyResults( extractValues(pivotResult))
                 .effectiveTimerange(pivotResult.effectiveTimerange())
                 .totalAggregatedMessages(pivotResult.total())
                 .sourceStreams(extractSourceStreams(streamsResult))
@@ -319,7 +319,12 @@ public class PivotAggregationSearch implements AggregationSearch {
                 }
             }
 
-            DateTime resultTimestamp = DateTime.parse(timeKey).withZone(DateTimeZone.UTC);
+            DateTime resultTimestamp;
+            try{
+                resultTimestamp = DateTime.parse(timeKey).withZone(DateTimeZone.UTC);
+            } catch (IllegalArgumentException e) {
+                throw new IllegalStateException("Failed to create event for: " + eventDefinition.title() + " (possibly due to non-existing grouping fields)", e);
+            }
             results.add(AggregationKeyResult.builder()
                     .key(groupKey)
                     .timestamp(resultTimestamp)

--- a/graylog2-server/src/test/java/org/graylog/events/processor/aggregation/AggregationEventProcessorTest.java
+++ b/graylog2-server/src/test/java/org/graylog/events/processor/aggregation/AggregationEventProcessorTest.java
@@ -104,7 +104,7 @@ public class AggregationEventProcessorTest {
     }
 
     @Test
-    public void testEventsFromAggregationResult() {
+    public void testEventsFromAggregationResult() throws EventProcessorException {
         final DateTime now = DateTime.now(DateTimeZone.UTC);
         final AbsoluteRange timerange = AbsoluteRange.create(now.minusHours(1), now.minusHours(1).plusMillis(SEARCH_WINDOW_MS));
 
@@ -195,7 +195,7 @@ public class AggregationEventProcessorTest {
     }
 
     @Test
-    public void testEventsFromAggregationResultWithConditions() {
+    public void testEventsFromAggregationResultWithConditions() throws EventProcessorException {
         final DateTime now = DateTime.now(DateTimeZone.UTC);
         final AbsoluteRange timerange = AbsoluteRange.create(now.minusHours(1), now.minusHours(1).plusMillis(SEARCH_WINDOW_MS));
 
@@ -403,7 +403,7 @@ public class AggregationEventProcessorTest {
     }
 
     @Test
-    public void testEventsFromAggregationResultWithEmptyResultUsesEventDefinitionStreamAsSourceStreams() {
+    public void testEventsFromAggregationResultWithEmptyResultUsesEventDefinitionStreamAsSourceStreams() throws EventProcessorException {
         final DateTime now = DateTime.now(DateTimeZone.UTC);
         final AbsoluteRange timerange = AbsoluteRange.create(now.minusHours(1), now.minusHours(1).plusMillis(SEARCH_WINDOW_MS));
 
@@ -447,7 +447,7 @@ public class AggregationEventProcessorTest {
     }
 
     @Test
-    public void testEventsFromAggregationResultWithEmptyResultAndNoConfiguredStreamsUsesAllStreamsAsSourceStreams() {
+    public void testEventsFromAggregationResultWithEmptyResultAndNoConfiguredStreamsUsesAllStreamsAsSourceStreams() throws EventProcessorException {
         final DateTime now = DateTime.now(DateTimeZone.UTC);
         final AbsoluteRange timerange = AbsoluteRange.create(now.minusHours(1), now.minusHours(1).plusMillis(SEARCH_WINDOW_MS));
 

